Daryun PR

dksekfbs72 opened the pull request:
# PR 생성 날짜
- 2022.12.19

# 주요 변경 사항
- 장바구니 담기/빼기 기능 구현
- 주문 결제/취소 기능 구현
- 주문 후 리뷰를 남기지 않으면 메일 발송

# 궁금한 점
- 리뷰 알림 메일을 전송할 때도 spring-batch를 적용하는게 좋을지 궁금합니다..!

dksekfbs72 committed:
util 패키지 추가, Distance 클래스 추가, CustomerService 분리

UserService와 CustomerService를 분리했습니다. Distance 클래스의 distance 메소드를 사용하면 두 좌표의 거리값을 double 타입으로 반환합니다. (단위 : km)

dksekfbs72 committed:
CustomerException 추가, 식당 리뷰/메뉴 보기, 장바구니 물건 담기 기능 구현

물건 담기는 Order Entity 의 menus 컬럼을 List<Integer> 로 변환하여 장바구니에 물건을 담을 때 해당 메뉴의 아이디를 add 하는 형식으로 구현했습니다.
OrderDto의 Response를 ListResponse/Response로 나누었습니다. 그냥 ListResponse에 하나만 넣어서 반환하는게 나은지 잘 모르겠습니다.

dksekfbs72 committed:
장바구니 담기 기능 오류 개선

장바구니 담기 기능 중 같은 식당의 메뉴인지 확인하는 구문에서 오류 수정했습니다.

dksekfbs72 committed:
주문 결제 기능 구현, 장바구니 메뉴 제거 기능 구현

주문 결제 기능은 실제 결제 기능은 구현되지 않았습니다.
enum OrderStatus 로 주문 상태를 표시합니다.

dksekfbs72 committed:
주문 후 일정 시간 후 리뷰 작성 메일 발송 구현, 주문 취소 구현

매 시간 주문 완료 후 리뷰를 달지 않은 고객에게 리뷰 등록 홍보 메일을 발송합니다.
주문 후 20분이 지나면 취소가 불가능합니다.

KimHyoJin commented on src/main/java/zerobase/demo/common/components/SchedulerComponents.java:
```diff
@@ -0,0 +1,45 @@
+	public void promotingReviewEmailSender() {
+		log.info("send email to promote review");
+
+		List<Order> orderList = orderRepository.findAllByOrderTimeBetweenAndReviewedAndStatus(
```
> 리뷰 알림 메일을 전송할 때도 spring-batch를 적용하는게 좋을지 궁금합니다..!

넵! 현재 findAllBy로 주문내역 전체를 가지고와서 리뷰를 써야하는 주문을 보여주고 있는데요, 만약 주문이 너무 많을 경우 너무 많은 주문내역이 한번에 메모리로 올라올 것 같아요. spring-batch를 쓴다면 chunk 단위로 처리가 가능하기 때문에 spring-batch를 이용하시는걸 추천드립니다

KimHyoJin commented on src/main/java/zerobase/demo/customer/service/Impl/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+
+	@Override
+	public boolean userAddReview(ReviewDto fromRequest, String userId) {
+		reviewRepository.findByOrderId(fromRequest.getOrderId())
```
review에서 한번 검색할 필요없이 order안의 reviewed flag로도 체크할수 있지 않을까요? 쿼리를 한번 줄일수 있을 것 같아요

dksekfbs72 replied:
아..! reviewed를 나중에 추가한거라 신경쓰지 못했네요..! 감사합니다 수정하겠습니다!

KimHyoJin commented on src/main/java/zerobase/demo/customer/controller/CustomerController.java:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	@GetMapping("/customer/{storeId}/menu")
+	public MenuDto.Response getStoreMenu(@PathVariable Integer storeId) {
```
사소한것이긴하지만, 보통 frontEnd와 협의할때는 형태를 동일하게 해서 내려줍니다. generic을 써서 Response를 만들고 조금 규격을 통일해보면 어떨까 싶습니다. 이런식으로 짜기도 합니다

~~~java
public class Response<T> {
    private String resultCode;
    private T result;
}
~~~

KimHyoJin commented on src/main/java/zerobase/demo/customer/service/Impl/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+	public OrderDto putShoppingBasket(Integer storeId, String username, Integer menuId, Integer count) {
+		Menu menu = menuRepository.findById(menuId).orElseThrow(
+			() -> new CustomerException(ResponseCode.MENU_NOT_FOUND));
+		if (menu.getSoldOut()) throw new CustomerException(ResponseCode.MENU_SOLD_OUT);
```
isSoldOut이라는 표현이 좀더 적합할 것 같아요.SoldOut이 Boolean이기때문에 isSoldOut도 자동생성 될거에요

KimHyoJin commented on src/main/java/zerobase/demo/customer/service/Impl/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+
+	@Override
+	public OrderDto cancelOrder(String username, Integer orderId) {
+		Order order = orderRepository.findById(orderId).orElseThrow(() -> new CustomerException(
```
payment가 완료된 order만 cancel될수 있도록 하는 로직이 추가되면 어떨까 싶어요

KimHyoJin commented on src/main/java/zerobase/demo/customer/service/Impl/CustomerServiceImpl.java:
```diff
@@ -0,0 +1,193 @@
+
+		List<Integer> newMenus = order.getMenus();
+		if (!newMenus.contains(menuId)) throw new CustomerException(ResponseCode.MENU_NOT_FOUND);
+		newMenus.remove(menuId);
```
장바구니에서 메뉴를 하나를 전체 삭제하는 걸까요? 장바구니에 A라는 메뉴가 3개 들어있을때 이 로직을 보면 A라는 메뉴 자체를 없애는 데 price는 1개에 대해서만 삭제하는 것 처럼 보여요. 이 API가 호출될때 장바구니에 든 A가 1개 삭제되는 것일까요 장바구니에 든 A가 전부 삭제되는 것일까요? 요거 제가 코드를 잘못 이해했을수도 있습니다!

dksekfbs72 replied:
아 장바구니 메뉴 삭제는 메뉴의 개수가 몇개든 한개씩만 삭제됩니다! A라는 메뉴가 3개면 삭제 API로 A메뉴의 ID값을 보내면 A메뉴가 2개가 되는 형식입니다..! 메뉴 삭제도 개수를 넣는 편이 좋을까요?

KimHyoJin reviewed (approved):
수고 많으셨어요! 간단히 몇가지 리뷰 드려보았습니다. 코드가 훨씬 깔끔해지셔서 리뷰하기 좋았습니다. 남은 기간 화이팅 해보아요!!!

dksekfbs72 committed:
PR 리뷰 사항 수정

UserController 와 CustomerController 의 반환 Response를 제네릭을 사용하여 규격 통일했습니다.
userAddReview 매소드의 예외 처리를 수정했습니다.
cancelOrder 매소드에 payment가 완료된 order만 cancel되도록 로직을 추가했습니다.
